torch/ao/quantization/qconfig_mapping_utils.py

```diff
@@ -0,0 +1,124 @@
+    for regex_pattern, qconfig in qconfig_mapping._module_name_regex_qconfig_dict.items():
+        if re.match(regex_pattern, module_name):
+            # first match wins
+            return qconfig
+    return fallback_qconfig
+
+
+def get_module_name_qconfig(qconfig_mapping, module_name, fallback_qconfig):
+    if module_name == '':
+        # module name qconfig not found
+        return fallback_qconfig
+    if module_name in qconfig_mapping._module_name_qconfig_dict:
+        return qconfig_mapping._module_name_qconfig_dict[module_name]
+    else:
+        parent, _ = _parent_name(module_name)
+        return get_module_name_qconfig(qconfig_mapping, parent, fallback_qconfig)
+
+
+def maybe_adjust_qconfig_for_module_type_or_name(qconfig_mapping, module_type, module_name, global_qconfig):
+    # get qconfig for module_name,
+    # fallback to module_name_regex_qconfig, module_type_qconfig,
+    # global_qconfig if necessary
+    module_type_qconfig = get_object_type_qconfig(
+        qconfig_mapping, module_type, global_qconfig)
+    module_name_regex_qconfig = get_module_name_regex_qconfig(
+        qconfig_mapping, module_name, module_type_qconfig)
+    module_name_qconfig = get_module_name_qconfig(
+        qconfig_mapping, module_name, module_name_regex_qconfig)
+    return module_name_qconfig
+
+
+def get_flattened_qconfig_dict(qconfig_mapping: QConfigMapping) -> Dict[Union[Callable, str], QConfigAny]:
+    """ flatten the global, object_type and module_name qconfig
+    to the same qconfig_dict so that it can be used by
+    propagate_qconfig_ function.
+    "module_name_regex" is ignored for now since it's not supported
+    in propagate_qconfig_, but it can be fixed later.
+
+    For example:
+    Input: {
+      "": qconfig,
+      "object_type": [
+        (torch.add, qconfig)
+      ],
+      "module_name": [
+        ("conv", qconfig)
+      ]
+    }
+
+    Output: {
+      "": qconfig,
+      torch.add: qconfig,
+      "conv": qconfig
+    }
+    """
+    flattened: Dict[Union[Callable, str], QConfigAny] = {"": qconfig_mapping.global_qconfig}
+    for obj, qconfig in qconfig_mapping._object_type_qconfig_dict.items():
+        flattened[obj] = qconfig
+    for obj, qconfig in qconfig_mapping._module_name_qconfig_dict.items():
+        flattened[obj] = qconfig
+    return flattened
+
+
+def convert_lists_to_ordered_dicts(qconfig_mapping: QConfigMapping):
+    """
+    Convert lists in a QConfigMapping to OrderedDicts.
+    """
+    qconfig_mapping._object_type_qconfig_dict = OrderedDict(
+        [(e.object_type, e.qconfig) for e in qconfig_mapping.object_type_qconfigs])
+    qconfig_mapping._module_name_qconfig_dict = OrderedDict(
+        [(e.module_name, e.qconfig) for e in qconfig_mapping.module_name_qconfigs])
+    qconfig_mapping._module_name_regex_qconfig_dict = OrderedDict(
+        [(e.module_name_regex, e.qconfig) for e in qconfig_mapping.module_name_regex_qconfigs])
```

Comment:
so these fields (e.g. _object_type_qconfig_dict) are filled only when we call this function, otherwise they would be empty, should we always update them while changing the QConfigMapping?

Reply:
not sure if this is exactly safe, how do we make sure we updates the internal dicts whenever there is an update? would it be cleaner if we don't make this internal?

Reply:
if we always regenerate when before we use it, I feel it might be cleaner that we just don't store this as a internal dict since when we store it, there will some confusion around how do we keep them in sync.

Reply:
I think we do need to keep these fields however because today we pass around `QConfigMapping` everywhere expecting ordered dicts. What we could do is remove these fields from the constructor, but then we would need to suppress mypy warnings everywhere we access the ordered dicts.

Reply:
I think we do need to keep these fields because today we pass around QConfigMapping everywhere expecting ordered dicts. What we could do is remove these fields from the constructor, but then we would need to suppress mypy warnings everywhere we access the ordered dicts.

Reply:
I think the semantics are fine if we always call `convert_lists_to_ordered_dicts(qconfig_mapping)` internally before each use. I also thought about freezing the lists after the populating the ordered dicts to prevent further changes, but I wasn't sure if the extra complexity is worth it.

Reply:
but calling `convert_lists_to_ordered_dicts` is not visible to QConfigMapping right? it's outside of the definition of QConfigMapping and nothing is guarantee that will be called, alternatively I feel we could also have a `build` method for QConfigMapping that will generate these internal dicts

Reply:
Ok, I was actually just preserving the existing behavior in this PR (which actually even replaced the lists using the same dict fields for minimum type safety), and this was quite difficult to follow. I like option (1), where we just maintain ordered dicts. I think this is the simplest.

Reply:
yeah just use ordered_dicts directly sounds good to me as well

Reply:
Sounds good. The new qconfig mapping uses ordered dicts directly. This also means we no longer need the dataclasses `QConfig*Entry`, which can't be used as dict keys because they are mutable.